Automattic/woocommerce-connect-client
# Change NPM with Yarn

Folks at Calypso tried this a few months ago but shelved it because Yarn wasn't mature enough. Relevant PR: https://github.com/Automattic/wp-calypso/pull/8660

Many big JS projects are already using it (including Webpack and React, for example), so I guess that's mature enough to give it a try.

Yarn is a JS package manager created by Facebook. It can be used as an (almost) drop-in replacement for NPM. Its advantages over NPM are:
* It has a locking mechanism built-in. Similar in concept to `shrinkwrap`, but it's used by default, so there's no way to install a package and forget to update `npm-shrinkwrap.json` or something like that.
* It's faster. Much faster. Expect installing `woocommerce-connect-client` from scratch to be twice as fast, or more.
* It mantains a local packages cache so it has some form of offline operation.
* It has a deterministic install alrogithm, as opposed to NPM3.
* Some cool new functionallity. For example: `yarn why packageName` will show why that package is installed (which dependency or sub-sub-subdependency is using it).
* It has more emojis.
* More info: https://yarnpkg.com/

To install Yarn: https://yarnpkg.com/en/docs/install#mac-tab

While I was at it, I re-arranged the `dependencies` and `devDependencies`, as we were being quite erratic deciding where to put each one. Since all the code is bundled, there are no real "dependencies" per-se, so I made an executive decision: **If something is needed to run `npm run i18n` and `npm run release`, then it's a `dependency`, else it's a `devDependency`**. That leaves as `devDependencies` testing-related stuff, linters and things like `webpack-dev-server`. This change can be back-ported to `master` even if we discard the rest of this PR.You can disagree with that, of course, as long as you provide another definition of `dependency` and `devDependency` :)

To test:
* Install Yarn.
* Run `yarn` (shortcut to `yarn install`).
* Check that everything works. The easiest way to do that is to run the tests and generate the release bundle, that should cover 99% of our code paths.
* To test that the `dependencies` are enough to build a release, remove `node_modules` and `yarn.lock`, then run `yarn install --production && npm run release` . It should work.

**Note**: In Windows, `yarn` has a race condition that affects the `karma` CLI. [More details](https://github.com/yarnpkg/yarn/issues/828#issuecomment-270035847). This shouldn't be a gating issue IMHO.
